File: src/dynamic-programming/maximum-subarray.ts
export function maxSubArray(nums: number[]): number {
  //dp[i]表示以nums[i]结尾最大的子序和dp[i]
  const dp: number[] = []

  dp[0] = nums[0]
  let ans = nums[0]

  for (let i = 1; i < nums.length; ++i) {
    if (dp[i - 1] < 0) {
      dp[i] = nums[i]
    } else {
      dp[i] = nums[i] + dp[i - 1]
    }

    ans = Math.max(ans, dp[i])
  }

  return ans
}

export function maxSubArray1(nums: number[]): number {
  //不难看出如果我们要求出最大子序和，那么只需要求出前缀和presum
  //然后在求出最大的presum[i] - presum[j]即可,其中j < i,  0<=i,j <= n
  const presum: number[] = []
  presum[0] = 0
  for (let i = 1; i <= nums.length; ++i) {
    presum[i] = presum[i - 1] + nums[i - 1]
  }

  let minPresum = 0
  let ans = nums[0]
  for (let i = 1; i < presum.length; ++i) {
    ans = Math.max(ans, presum[i] - minPresum)
    minPresum = Math.min(presum[i], minPresum)
  }

  return ans
}

File: src/dynamic-programming/maximum-subarray.c
int maxSubArray(int *nums, int numsSize)
{
    int *dp = malloc(sizeof(int) * numsSize);

    int ans = nums[0];
    dp[0] = nums[0];
    for (int i = 1; i < numsSize; ++i)
    {
        dp[i] = nums[i];

        if (dp[i - 1] + nums[i] > nums[i])
        {
            dp[i] = dp[i - 1] + nums[i];
        }

        ans = ans > dp[i] ? ans : dp[i];
    }

    return ans;
}

int maxSubArray(int *nums, int numsSize)
{
    int *presum = malloc(sizeof(int) * (numsSize + 1));

    presum[0] = 0;
    for (int i = 1; i <= numsSize; ++i)
    {
        presum[i] = presum[i - 1] + nums[i - 1];
    }

    int presumMin = 0;
    int ans = nums[0];
    for (int i = 1; i <= numsSize; ++i)
    {
        ans = presum[i] - presumMin > ans ? presum[i] - presumMin : ans;

        presumMin = presum[i] < presumMin ? presum[i] : presumMin;
    }

    return ans;
}